### PR TITLE
[ui] Graph Editor Update: Quick Node Coloring with the Color Selector Tool

### PR DESCRIPTION
**Pull request by waaake**
## Features list
* The Color Selector Provides a really quick way to apply color onto selected nodes or to remove it.
* Currently holds a limited set of frequently used colors.
* Uses 'C' Key as the shortcut to toggle the Color Selector.

### CODE REVIEW

**Review by yann-lty — CHANGES_REQUESTED** [review]
Implementation-wise: this should rely on the existing SetAttributeCommand.
UX-wise: I would suggest re-working the color palette for better contrast with the font color.

File: meshroom/ui/qml/Controls/ColorSelector.qml [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+ */
+MaterialToolButton {
+    id: root
+
+    text: MaterialIcons.palette
+
+    // Internal property holding when the popup remains visible and when is it toggled off
+    property var isVisible: false
+
+    property var colors: [
```
[the remark this answers is not in the document]
Comment (natowi) [reply meshroom/ui/qml/Controls/ColorSelector.qml:19]: Better color just the header like in blender:
grafik

[the remark this answers is not in the document]
Comment (fabiencastan) [reply meshroom/ui/qml/Controls/ColorSelector.qml:19]: It's aligned with Blender, the title bar as a type color, and the color of the bottom part could be customized:
image

[the remark this answers is not in the document]
Comment (waaake) [reply meshroom/ui/qml/Controls/ColorSelector.qml:19]: @yann-lty , thanks for the feedback. I do agree that the colours do not play well with regards to constrast. For now, I think we can definitely go with darker shades.
But do you think the colour of the text should react with the background colour ?
What I mean is if the colour has a lighter shade then the text colour should be black or darker, and vice-versa. 
We can do this implementation later (if you think that's required)

[the remark this answers is not in the document]
Comment (waaake) [reply meshroom/ui/qml/Controls/ColorSelector.qml:19]: Screenshot 2024-11-27 at 19 05 57

Now we have Darker shades 🙂

File: meshroom/core/node.py [hunk 2]
```diff
@@ -556,6 +556,15 @@ def getColor(self):
             return self.internalAttribute("color").value.strip()
         return ""
 
+    def setColor(self, color: str):
```
Comment (yann-lty) [inline meshroom/core/node.py:559]: With regards to what this PR is addressing, adding the setter for this property is not needed.

File: meshroom/core/node.py [hunk 3]
```diff
@@ -1386,7 +1395,7 @@ def has3DOutputAttribute(self):
     internalAttributes = Property(BaseObject, getInternalAttributes, constant=True)
     internalAttributesChanged = Signal()
     label = Property(str, getLabel, notify=internalAttributesChanged)
-    color = Property(str, getColor, notify=internalAttributesChanged)
+    color = Property(str, getColor, setColor, notify=internalAttributesChanged)
```
Comment (yann-lty) [inline meshroom/core/node.py:1398]: (same as above) Adding the setter is not needed in this scope.

File: meshroom/ui/graph.py [hunk 4]
```diff
@@ -952,6 +952,20 @@ def selectFollowing(self, node):
...
 
+    @Slot(str)
+    def setSelectedNodesColor(self, color):
+        """ Sets the Provided color on the selected Nodes.
+
+        Args:
+            color (str): Hex code of the color to be set on the nodes.
+        """
+        # Update the color attribute of the nodes which are currently selected
+        with self.groupedGraphModification("Update Color for Select Nodes"):
```
Comment (yann-lty) [inline meshroom/ui/graph.py:963]: For the command name that appears in the undo stack:

```suggestion
        with self.groupedGraphModification("Set Nodes Color"):
```

I think we could go with removing the notion of "selected" in those, as selection operations are not undoable - therefore, undoing "Set Selected Nodes Color" is a bit ambiguous if the selection has changed in between.
What do you think?

File: meshroom/ui/qml/Controls/ColorSelector.qml (hunk 1 above)
[the remark this answers is not in the document]
Comment (yann-lty) [reply meshroom/ui/qml/Controls/ColorSelector.qml:19]: Thanks for the update!  
For the previous question: this is working well for a first iteration, we can think of a more advanced system in a 2nd phase if needed.

**Review by yann-lty — CHANGES_REQUESTED** [review]
Minor fixes left, but LGTM after that 👍.

File: meshroom/ui/graph.py (hunk 4 above)
Comment (waaake) [reply meshroom/ui/graph.py]: Yes, makes sense, I probably did not notice this while working on the previous comments. Thanks for the suggestion.

File: meshroom/core/node.py (hunk 3 above)
Comment (waaake) [reply meshroom/core/node.py]: This has been removed.

**Review by yann-lty — APPROVED** [review]
LGTM

**Review by yann-lty — CHANGES_REQUESTED** [review]
The "C" shortcut is currently breaking the Copy action (Ctrl+C).

File: meshroom/ui/qml/GraphEditor/GraphEditor.qml [hunk 5]
```diff
@@ -136,6 +136,8 @@ Item {
     Keys.onPressed: function(event) {
         if (event.key === Qt.Key_F) {
             fit()
+        } else if (event.key === Qt.Key_C) {
```
[the remark this answers is not in the document]
Comment (waaake) [reply meshroom/ui/qml/GraphEditor/GraphEditor.qml]: Yeah, it's my bad, I did not notice something as simple as this. Have updated the code to consider both cases.

**Review by yann-lty — APPROVED** [review]
:+1:

**Pull request merged**